src/coreclr/jit/llvm.cpp

```diff
@@ -759,62 +806,156 @@ Value* buildCall(llvm::IRBuilder<>& builder, GenTree* node)
             // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
             (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
 
-            return mapGenTreeToValue(node, builder.CreateCall(llvmFunc, getShadowStackForCallee(builder)));
+            mapGenTreeToValue(node, builder.CreateCall(llvmFunc, getShadowStackForCallee(builder)));
+            return;
         }
     }
     else if (call->gtCallType == CT_USER_FUNC && !call->IsVirtualStub() /* TODO: Virtual stub not implemented */)
     {
-        return buildUserFuncCall(call, builder);
+        buildUserFuncCall(call, builder);
+        return;
     }
     failFunctionCompilation();
 }
 
-Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
+void buildCast(llvm::IRBuilder<>& builder, GenTreeCast* cast)
+{
+    if (cast->gtCastType == TYP_BOOL && cast->TypeGet() == TYP_INT && cast->gtGetOp1()->TypeGet() == TYP_INT)
+    {
+        Value* intValue = builder.CreateZExt(getGenTreeValue(cast->gtGetOp1()).getValue(builder), getLLVMTypeForVarType(TYP_INT));
+        mapGenTreeToValue(cast, intValue); // nothing to do except map the source value to the destination GenTree
```

Comment:
```suggestion
    if ((cast->CastToType() == TYP_BOOL) && cast->TypeIs(TYP_INT) && cast->CastOp()->TypeIs(TYP_INT))
    {
        Value* intValue = builder.CreateZExt(getGenTreeValue(cast->CastOp()).getValue(builder), getLLVMTypeForVarType(TYP_INT));
        mapGenTreeToValue(cast, intValue); // nothing to do except map the source value to the destination GenTree
```

So when we call `getGenTreeValue(cast->CastOp()).getValue(builder)`, what type does it have? I would expect `int`, but then documentation states that we must use a smaller type for `zext` source than the destination... So how does it really work here?

(This all comes down to `CAST(BOOL)` actually being `int -> bool -> int`, and similar for all other small types)

Reply:
An example
```
------------ BB02 [000..007) (return), preds={BB01} succs={}
               [000010] ------------                 IL_OFFSET void   IL offset: 0x0
N002 (  3,  2) [000001] ------------         t1 =    LCL_VAR   ref    V00 this         u:1 (last use)
                                                  /--*  t1     ref    this in rcx
N003 ( 17,  6) [000002] --CXG-------         t2 = *  CALL      int    System.Type.get_IsSignatureType
                                                  /--*  t2     int
N004 ( 18,  8) [000003] ---XG-------         t3 = *  CAST      int <- bool <- int
                                                  /--*  t3     int
N006 ( 22, 11) [000005] DA-XG-------              *  STORE_LCL_VAR int    V02 tmp1         d:1
               [000011] ------------                 IL_OFFSET void   IL offset: 0x6
N001 (  3,  2) [000006] ------------         t6 =    LCL_VAR   int    V02 tmp1         u:1 (last use)
                                                  /--*  t6     int
N002 (  4,  4) [000009] ------------         t9 = *  CAST      int <- bool <- int
                                                  /--*  t9     int
N003 (  5,  5) [000007] ------------              *  RETURN    int
```
`CastOp` :
![image](https://user-images.githubusercontent.com/2720041/134988115-03e8148f-e045-4188-8670-60b74b0573a4.png)
And it gets from `getGenTreeValue` an `i8`.
![image](https://user-images.githubusercontent.com/2720041/134988374-75527b42-33ef-47e6-b1e0-4a86e72f53fa.png)

When creating the call the `CORINFO_SIG_INFO` is
![image](https://user-images.githubusercontent.com/2720041/134988734-5c85b517-9e0d-4510-9530-02c6f0da9e52.png)
 so we end up ok, an i8 to i32
```
  %6 = call i8 @S_P_CoreLib_System_Type__get_IsSignatureType(i8* %4), !dbg !3637
  %7 = zext i8 %6 to i32, !dbg !3637
```

Reply:
I see.

> All the example I've seen are from a call, so maybe just getting lucky so far.

I would say yes. I suppose `bool`s are special because they do not actually exist. One other source of casts with them is "normalization on store", but that is harder to hit because C# only allows boolean expressions on the RHS, and those do not need normalization. Anyhow, this will need to be tackled in any case when other casts are implemented.